File: software/host/py/host.py
import sys
import codecs
hexify = codecs.getencoder('hex')
import traceback

import logging
#logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
#logging.basicConfig(stream=sys.stdout, level=logging.WARNING)
logging.basicConfig(
            level=logging.DEBUG,
            format="%(asctime)s %(name)s %(levelname)-8s %(thread)d %(message)s",
            datefmt="%Y-%m-%d %H:%M:%S")

logger = logging.getLogger("bramnik")

from smbus2 import SMBus, i2c_msg, SMBusWrapper
#from smbus import SMBus
bus = SMBus(1) # 0 indicates /dev/i2c-0

import time
import RPi.GPIO as GPIO




# I2C consts
address = 0x68
offset = 0

# Constants
NORMAL_DELAY = 1
EXCEPTION_DELAY = 2
# commands
PLAY_GRANTED = 0x20
PLAY_DENIED = 0x21
STATUS = 0x30
READ_NFC = 0x31
READ_KEYPAD = 0x32

# HW consts
DOOR_GPIO_PIN = 17  #BCM.17

GPIO.setmode(GPIO.BCM)
GPIO.setup(DOOR_GPIO_PIN, GPIO.OUT)


# I2C low level
def read(num):
    #with SMBusWrapper(1) as bus:
    #msg = [bus.read_byte(address) for x in range(0,num)]
    msg = i2c_msg.read(address, num)
    bus.i2c_rdwr(msg)
    return list(msg)

def write(command):
    #with SMBusWrapper(1) as bus:
    bus.write_byte(address, command)
def open_door():
    logger.warning("Openning door")
    GPIO.output(DOOR_GPIO_PIN, 1)
    time.sleep(0.1)
    GPIO.output(DOOR_GPIO_PIN, 0)


# logic
def read_status():
    #return bytearray([1])
    write(STATUS)
    time.sleep(0.1)
    s = read(1)
    return s

# reads NFC and checks if it is valid
def check_nfc():
    logger.debug("checking nfc")
    write(READ_NFC)
    time.sleep(0.1)
    s = read(16)
    logger.warning("checking nfc: %s", s)
    if s[:3] == [131, 78, 213]:
        open_door()
        write(PLAY_GRANTED)
    else:
        logger.warning("access denied for: %s", s)
        write(PLAY_DENIED)

# reads code and checks if it is valid
def check_code():
    logger.debug("checking code")
    write(READ_KEYPAD)
    time.sleep(0.1)
    s = read(16)
    logger.warning("checking code: %s", s)
    if s[:4] == [49,50,51,52]: # 1,2,3,4 on keypad
        open_door()
        write(PLAY_GRANTED)
    else:
        logger.warning("access denied for: %s", s)
        write(PLAY_DENIED)

def main_loop():
    logger.warning("starting main loop")
    delay = NORMAL_DELAY
    while(True):
        try:
            logger.debug("loop")
            s = read_status()
            s = s[0]
            logger.info("status: %s", s)
            if s&1: # lower bit means reader has NFC data
                check_nfc()
            if s&2: # second bit means reader has code data
                check_code()

        except Exception as e:
            logger.error("Exception: %s", e)
            logger.error(traceback.format_exc())
            delay = EXCEPTION_DELAY
        else:
            delay = NORMAL_DELAY
        #delay
        time.sleep(delay)




main_loop()

File: obsolete/software/host/host.rb
require 'i2c'
require 'logger'
require 'base64'
require_relative 'user'

class Bramnik
  attr_accessor :logger
  def initialize(address = 0x68, device = '/dev/i2c-1', logfile = '/tmp/bramnik.log')
    @address = address
    @i2c = ::I2C.create(device)
    `gpio mode 0 out`
    @logger = Logger.new(logfile)
  end

  def read(n = 1)
    data = @i2c.read(@address, n)
    @logger.info "Received: #{Base64.encode64 data}" unless data == "\x00"
    data
  end

  def write(data)
    @logger.info "Sent: #{data}" unless data == 48
    @i2c.write(@address, data)
  end

  def status
    write(0x30)
    read 1
  end

  def read_nfc
    write 0x31
    data = Base64.encode64 read 32
    p data
    if User.check_access data
      play_granted
      open_door
    else
      play_denied
    end
    data
  end

  def read_keyboard
    write 0x32
    data = Base64.encode64 read 32
    p data
    if User.check_code data
      play_granted
      open_door
    else
      play_denied
    end
    data
  end

  def disable
    write 0x10
  end

  def enable_reader_disable_keypad
    write 0x11
  end

  def disable_reader_enable_keypad
    write 0x12
  end

  def enable
    write 0x13
  end

  def play_granted
    write 0x21
  end

  def play_denied
    write 0x20
  end

  def open_door
    `gpio write 0 1`
    sleep 0.2
    `gpio write 0 0`
  end
end

bramnik = Bramnik.new

while true
  status = bramnik.status
  case status
    when "\x00"
    when "\x01"
      bramnik.read_nfc
    when "\x02"
      bramnik.read_keyboard
    else
      bramnik.logger.warn "Wrong status #{status}"
  end
  sleep 1
end